File: pkg/dart2cpp/test_complex_header.dart
/// 复杂示例，测试头文件生成功能
class Vector2D {
  double x;
  double y;
  
  Vector2D(this.x, this.y);
  
  double magnitude() {
    var sum = x * x + y * y;
    return sum.sqrt();
  }
  
  Vector2D operator+(Vector2D other) {
    return Vector2D(x + other.x, y + other.y);
  }
}

/// 数学扩展
extension MathExtension on double {
  double sqrt() {
    if (this < 0) return 0;
    double x = this;
    double prev = 0;
    while ((x - prev).abs() > 0.0001) {
      prev = x;
      x = (x + this / x) / 2;
    }
    return x;
  }
}

/// 泛型扩展
extension LetExtension<T> on T {
  R let<R>(R Function(T) block) => block(this);
}

/// 泛型类
class Container<T> {
  T value;
  
  Container(this.value);
  
  T getValue() => value;
}

/// 泛型函数
T identity<T>(T value) {
  return value;
}

void main() {
  var v1 = Vector2D(3, 4);
  print('Magnitude: ${v1.magnitude()}');
  
  var result = v1.let((v) => v.magnitude());
  print('Let result: $result');
  
  var container = Container<Vector2D>(v1);
  print('Container value magnitude: ${container.getValue().magnitude()}');
}

File: pkg/dart2cpp/test_complex_header.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: Vector2D
// ============================================================================

class Vector2D {
public:
  Double x;
  Double y;
  Vector2D(Double x, Double y) : x(x), y(y) {
  }
  
  Double magnitude() {
    auto sum = this->x->operator_mul(this->x)->operator_add(this->y->operator_mul(this->y));
return MathExtension::sqrt(sum);
  }
  
  ObjectPtr<Vector2D> operator_add(ObjectPtr<Vector2D> other) {
    return ObjectPtr<Vector2D>(new Vector2D(this->x->operator_add(other->x), this->y->operator_add(other->y)));
  }
  
};

// ============================================================================
// 类: Container
// ============================================================================

template<typename T>
class Container {
public:
  T value;
  Container(T value) : value(value) {
  }
  
  T getValue() {
    return this->value;
  }
  
};

// ============================================================================
// Extension: MathExtension
// ============================================================================

namespace MathExtension {
  inline Double sqrt(Double this_) {
    if (this_->operator_less(dart_int(0))) {
return dart_double(0.0);
}
auto x = this_;
auto prev = dart_double(0.0);
while (x->operator_sub(prev)->abs()->operator_greater(dart_double(0.0001))) {
prev = x;
x = x->operator_add(this_->operator_div(x))->operator_div(dart_double(2.0));
}
return x;
  }
  
} // namespace MathExtension

// ============================================================================
// Extension: LetExtension
// ============================================================================

namespace LetExtension {
  template<typename T, typename R>
  inline R let(T this_, ObjectPtr<TypedFunction<std::function<R(T)>, R, T>> block) {
    return block->call(this_);
  }
  
} // namespace LetExtension

template<typename T>
T identity(T value);
Double MathExtension::sqrt(Double this_);
template<typename T, typename R>
R LetExtension::let(T this_, ObjectPtr<TypedFunction<std::function<R(T)>, R, T>> block);
template<typename T>
T identity(T value) {
  return value;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto v1 = ObjectPtr<Vector2D>(new Vector2D(dart_double(3.0), dart_double(4.0)));
dart_print(dart_string("Magnitude: ") + (v1->magnitude()).toString());
auto result = LetExtension::let(v1, makeFunction([&](ObjectPtr<Vector2D> v) { return v->magnitude(); }));
dart_print(dart_string("Let result: ") + (result).toString());
auto container = ObjectPtr<Container<ObjectPtr<Vector2D>>>(new Container<ObjectPtr<Vector2D>>(v1));
dart_print(dart_string("Container value magnitude: ") + (container->getValue()->magnitude()).toString());
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
